bindings/python/manifold3d.cpp

```diff
@@ -22,107 +22,180 @@
-  for (int i = 0; i < nvert; i++) {
-    for (int j = 0; j < N; j++) {
-      buffer[i * N + j] = begin[i][j];
+  bool from_python(handle src, uint8_t flags, cleanup_list *cleanup) noexcept {
+    int rows = PyObject_Size(src.ptr());  // negative on failure
+    if (rows != R) return false;
+    for (size_t i = 0; i < R; i++) {
+      const nb::object &slice = src[i];
+      int cols = PyObject_Size(slice.ptr());  // negative on failure
+      if (cols != C) return false;
+      for (size_t j = 0; j < C; j++) {
+        make_caster<T> t_cast;
+        if (!t_cast.from_python(slice[j], flags, cleanup)) return false;
+        value[j][i] = t_cast.value;
+      }
     }
+    return true;
   }
-  return {buffer, {nvert, N}, mem_mgr};
-}
+  static handle from_cpp(glm_type mat, rv_policy policy,
+                         cleanup_list *cleanup) noexcept {
+    T *buffer = new T[R * C];
+    nb::capsule mem_mgr(buffer, [](void *p) noexcept { delete[](T *) p; });
+    for (int i = 0; i < R; i++) {
+      for (int j = 0; j < C; j++) {
+        // py is (Rows, Cols), glm is (Cols, Rows)
+        buffer[i * C + j] = mat[j][i];
+      }
+    }
+    numpy_type arr{buffer, {R, C}, std::move(mem_mgr)};
+    return ndarray_wrap(arr.handle(), int(ndarray_framework::numpy), policy,
+                        cleanup);
+  }
+};
 
-// helper to convert std::vector<glm::vec*> to numpy
-template <class T, int N, glm::qualifier Precision>
-nb::ndarray<nb::numpy, T, nb::shape<nb::any, N>> to_numpy(
-    std::vector<glm::vec<N, T, Precision>> const &vec) {
-  return to_numpy(vec.data(), vec.data() + vec.size());
-}
+// handle std::vector<glm::vecN>
+template <class T, int N, glm::qualifier Q>
+struct nb::detail::type_caster<std::vector<glm::vec<N, T, Q>>> {
+  using glm_type = glm::vec<N, T, Q>;
+  using numpy_type = nb::ndarray<nb::numpy, T, nb::shape<nb::any, N>>;
+  NB_TYPE_CASTER(std::vector<glm_type>,
+                 const_name(glm_name<glm_type>::multi_name));
 
-// helper to convert numpy to std::vector<glm::vec*>
-template <class T, size_t N, glm::qualifier Precision = glm::defaultp>
-void to_glm_range(nb::ndarray<nb::numpy, T, nb::shape<nb::any, N>> const &arr,
-                  glm::vec<int(N), T, Precision> *begin,
-                  glm::vec<int(N), T, Precision> *end) {
-  if (arr.shape(0) != end - begin) {
-    throw std::runtime_error(
-        "received numpy.shape[0]: " + std::to_string(arr.shape(0)) +
-        " expected: " + std::to_string(int(end - begin)));
+  bool from_python(handle src, uint8_t flags, cleanup_list *cleanup) noexcept {
+    make_caster<numpy_type> arr_cast;
+    if (arr_cast.from_python(src, flags, cleanup)) {
+      int num_vec = arr_cast.value.shape(0);
+      value.resize(num_vec);
+      for (int i = 0; i < num_vec; i++) {
+        for (int j = 0; j < N; j++) {
+          value[i][j] = arr_cast.value(i, j);
+        }
+      }
+    } else {
+      int num_vec = PyObject_Size(src.ptr());  // negative on failure
```

Comment:
This is surprising. Just curious, any reference on this?

Reply:
So this function is `noexcept` and if you call `nb::len` on a `float` it throws, and terminates the program. So I looked into source for `nb::len` and found this function which gives a softer fallback mode.